README.rst
=============================
django-restricted-sessions
=============================

.. image:: https://badge.fury.io/py/django-restricted-sessions.png
    :target: http://badge.fury.io/py/django-restricted-sessions

.. image:: https://travis-ci.org/mxsasha/django-restricted-sessions.png?branch=master
    :target: https://travis-ci.org/mxsasha/django-restricted-sessions

.. image:: https://coveralls.io/repos/mxsasha/django-restricted-sessions/badge.png?branch=master&
    :target: https://coveralls.io/r/mxsasha/django-restricted-sessions?branch=master

Restricts Django sessions to IP and/or user agent.

If the IP or user agent changes after creating the session, the a 400 response is given to the request, the session is
flushed (all session data deleted, new session created) and a warning is logged. The goal of this middleware is to
make it harder for an attacker to use a session ID they obtained. It does not make abuse of session IDs impossible.

For compatibility with IPv6 privacy extensions, by default only the first 64 bits of an IPv6 address are checked.

Documentation
-------------

The full documentation is at https://django-restricted-sessions.readthedocs.org.

Quickstart
----------

Install django-restricted-sessions::

    pip install django-restricted-sessions

Then add it to your middleware after SessionMiddleware::

    MIDDLEWARE = [
        "django.middleware.security.SecurityMiddleware",
        "django.contrib.sessions.middleware.SessionMiddleware",
        'restrictedsessions.middleware.RestrictedSessionsMiddleware',
        ....
    ]

If you use ``RESTRICTEDSESSIONS_AUTHED_ONLY``, ensure this middleware is added after
``AuthenticationMiddleware`` so that the ``request.user`` is present.


HISTORY.rst
.. :changelog:

History
-------

0.4.0 (2023-06-02)
++++++++++++++++++

* Updated Django support to only 3.2, 4.1 and 4.2 (thanks to @ronnievdc).
* Various other cleanups and documentation format fixes.

0.3.0 (2019-12-03)
++++++++++++++++++

* Updated Django support to 1.11, 2.2 and 3.0.
* Fixed issues when ``X_FORWARDED_FOR`` contains multiple addresses.

0.2.0 (2017-04-06)
++++++++++++++++++

* For Django 1.10+ support, changed from `object` to `django.utils.deprecation.MiddlewareMixin`
* Added PyPI trove classifiers for Django versions and more Python versions
* Updated `travis.yml` for more Python versions

0.1.4 (2016-07-02)
++++++++++++++++++

* Fixed an exception that could occur when non-utf8 bytes were included
  in user agent strings.

0.1.3.1 (2016-05-26)
++++++++++++++++++++

* Version bump to avoid PyPI's duplicate filename ban.

0.1.3 (2016-05-26)
++++++++++++++++++

* Added support to redirect to known view, or use custom status code settings.
* Added support for ignoring unauthenticated sessions.
* Fixed short circuit when REMOTE_ADDR was unknown.
* Dropped support for older Python versions: now requires 2.7, 3.3 or newer,
  with Django 1.8.

0.1.2 (2014-03-20)
++++++++++++++++++

* Resolved exception being raised when session switches from IPv4 to IPv6
* Python 3.4 support

0.1.1 (2014-02-18)
++++++++++++++++++

* Added missing netaddr requirement to setup.py.

0.1.0 (2014-02-17)
++++++++++++++++++

* First release on PyPI.


test_urls.py
from django.urls import re_path
from django.http import HttpResponse


urlpatterns = [
    re_path(
        r'test_view/',
        view=lambda r: HttpResponse(content="For unittests", status=200),
        name='test_view'
    ),
]


tests/test_middleware.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_django-restricted-sessions
------------

Tests for `django-restricted-sessions` models module.
"""

from __future__ import unicode_literals

from unittest.mock import Mock

from django.contrib.auth.models import User, AnonymousUser
from django.contrib.sessions.middleware import SessionMiddleware
from django.http import HttpResponse
from django.test import TestCase
from django.test.client import RequestFactory
from django.test.utils import override_settings

from restrictedsessions import middleware


class TestRestrictedsessionsMiddleware(TestCase):
    def setUp(self):
        self.get_response = Mock()
        self.middleware = middleware.RestrictedSessionsMiddleware(self.get_response)
        self.factory = RequestFactory()
        self.request = self.factory.get("/")

    def test_without_session(self):
        self.assertTrue(self.middleware.process_request(self.request) is None)

    def test_without_remote_addr(self):
        self.add_session_to_request()
        del self.request.META["REMOTE_ADDR"]
        self.assertTrue(self.middleware.process_request(self.request) is None)
        self.assertTrue(self.request.session.get(middleware.SESSION_IP_KEY) is None)
        self.assertTrue(self.request.session.get(middleware.SESSION_UA_KEY) is None)

    def test_without_remote_addr_still_check_user_agent(self):
        # Given: No remote IP incoming
        self.add_session_to_request()
        del self.request.META["REMOTE_ADDR"]
        # Given: First request has an arbitrary User Agent string
        initial_user_agent = "ua-initial"
        self.request.META["HTTP_USER_AGENT"] = initial_user_agent
        # When: the middleware first sees the request
        self.assertTrue(self.middleware.process_request(self.request) is None)
        # Then: it sets only the user agent string
        self.assertTrue(self.request.session.get(middleware.SESSION_IP_KEY) is None)
        self.assertEqual(self.request.session.get(middleware.SESSION_UA_KEY), initial_user_agent)

        # Given: A second request is made with a new user agent string
        different_user_agent = "us-changed"
        self.request.META["HTTP_USER_AGENT"] = different_user_agent
        # When: the middleware first sees the request
        response = self.middleware.process_request(self.request)
        # Then: there was an HttpResponse returned from middleware
        self.assertIsInstance(response, HttpResponse)
        # Then: the response was a HTTP 400 status response by default
        self.assertEqual(response.status_code, 400)

    @override_settings(RESTRICTEDSESSIONS_FAILURE_STATUS=404)
    def test_without_remote_addr_still_check_user_agent_when_configured_status(self):
        # Given: No remote IP incoming
        self.add_session_to_request()
        del self.request.META["REMOTE_ADDR"]
        # Given: First request has an arbitrary User Agent string
        initial_user_agent = "ua-initial"
        self.request.META["HTTP_USER_AGENT"] = initial_user_agent
        # When: the middleware first sees the request
        self.assertTrue(self.middleware.process_request(self.request) is None)
        # Then: it sets only the user agent string
        self.assertTrue(self.request.session.get(middleware.SESSION_IP_KEY) is None)
        self.assertEqual(self.request.session.get(middleware.SESSION_UA_KEY), initial_user_agent)

        # Given: A second request is made with a new user agent string
        different_user_agent = "us-changed"
        self.request.META["HTTP_USER_AGENT"] = different_user_agent
        # When: the middleware first sees the request
        response = self.middleware.process_request(self.request)
        # Then: there was an HttpResponse returned from middleware
        self.assertIsInstance(response, HttpResponse)
        # Then: the response was a HTTP 400 status response by default
        self.assertEqual(response.status_code, 404)

    @override_settings(RESTRICTEDSESSIONS_REDIRECT_VIEW="test_view")
    def test_without_remote_addr_still_check_user_agent_when_redirect(self):
        # Given: No remote IP incoming
        self.add_session_to_request()
        del self.request.META["REMOTE_ADDR"]
        # Given: First request has an arbitrary User Agent string
        initial_user_agent = "ua-initial"
        self.request.META["HTTP_USER_AGENT"] = initial_user_agent
        # When: the middleware first sees the request
        self.assertTrue(self.middleware.process_request(self.request) is None)
        # Then: it sets only the user agent string
        self.assertTrue(self.request.session.get(middleware.SESSION_IP_KEY) is None)
        self.assertEqual(self.request.session.get(middleware.SESSION_UA_KEY), initial_user_agent)

        # Given: A second request is made with a new user agent string
        different_user_agent = "us-changed"
        self.request.META["HTTP_USER_AGENT"] = different_user_agent
        # When: the middleware first sees the request
        response = self.middleware.process_request(self.request)
        # Then: there was an HttpResponse returned from middleware
        self.assertIsInstance(response, HttpResponse)
        # Then: the response was an HTTP redirect to the test view
        self.assertEqual(response.status_code, 302)
        # Then: the URL is the view from the settings override
        self.assertEqual(response.url, "/test_view/")

    @override_settings(RESTRICTEDSESSIONS_REMOTE_ADDR_KEY="CUSTOM_REMOTE_ADDR")
    def test_without_remote_addr_with_custom_key(self):
        self.add_session_to_request()
        self.assertTrue(self.middleware.process_request(self.request) is None)
        self.assertTrue(self.request.session.get(middleware.SESSION_IP_KEY) is None)
        self.assertTrue(self.request.session.get(middleware.SESSION_UA_KEY) is None)

    def test_saves_with_remote_addr(self):
        self.add_session_to_request()
        self.assertTrue(self.middleware.process_request(self.request) is None)
        self.assertEqual(
            self.request.session[middleware.SESSION_IP_KEY], self.request.META["REMOTE_ADDR"]
        )
        self.request.META["HTTP_USER_AGENT"] = "test-ua"
        self.assertTrue(self.middleware.process_request(self.request) is None)
        self.assertEqual(
            self.request.session[middleware.SESSION_IP_KEY], self.request.META["REMOTE_ADDR"]
        )
        self.assertEqual(
            self.request.session[middleware.SESSION_UA_KEY], self.request.META["HTTP_USER_AGENT"]
        )

    @override_settings(RESTRICTEDSESSIONS_REMOTE_ADDR_KEY="CUSTOM_REMOTE_ADDR")
    def test_saves_with_remote_addr_with_custom_key(self):
        self.add_session_to_request()
        self.request.META["CUSTOM_REMOTE_ADDR"] = self.request.META["REMOTE_ADDR"]
        del self.request.META["REMOTE_ADDR"]
        self.assertTrue(self.middleware.process_request(self.request) is None)
        self.assertEqual(
            self.request.session[middleware.SESSION_IP_KEY], self.request.META["CUSTOM_REMOTE_ADDR"]
        )

    def test_validates_ipv4(self):
        self._remote_addr_test(session_ip="127.0.0.1", valid="127.0.0.1", invalid="127.0.0.2")

    @override_settings(RESTRICTEDSESSIONS_IPV4_LENGTH=24)
    def test_validates_ipv4_subnet(self):
        self._remote_addr_test(session_ip="127.0.0.1", valid="127.0.0.255", invalid="127.0.1.1")

    def test_validates_ipv6(self):
        self._remote_addr_test(
            session_ip="2001:db8:0:0:0:0:0:1",
            valid="2001:db8:0:0:ffff:ffff:ffff:ffff",
            invalid="2001:db8:0:1:0:0:0:1",
        )

    @override_settings(RESTRICTEDSESSIONS_IPV6_LENGTH=128)
    def test_validates_ipv6_custom_length(self):
        self._remote_addr_test(
            session_ip="2001:db8:0:0:0:0:0:1",
            valid="2001:db8:0:0:0:0:0:1",
            invalid="2001:db8:0:0:0:0:0:2",
        )

    def test_validates_ipv6_mapped_ipv4(self):
        self._remote_addr_test(
            session_ip="::ffff:127.0.0.1", valid="::ffff:127.0.0.1", invalid="::ffff:127.0.0.2"
        )

    def test_validates_multiple_addresses(self):
        self._remote_addr_test(
            session_ip="127.0.0.1", valid="127.0.0.1, 192.0.2.1", invalid="127.0.0.2, 127.0.0.1"
        )

    def test_validates_ipv4_to_ipv6(self):
        self._remote_addr_test(session_ip="127.0.0.1", valid="127.0.0.1", invalid="2001:db8::1")

    def test_validates_ipv6_to_ipv4(self):
        self._remote_addr_test(session_ip="2001:db8::1", valid="2001:db8::1", invalid="127.0.0.1")

    def _remote_addr_test(self, session_ip, valid, invalid=None):
        self.add_session_to_request()
        self.request.session[middleware.SESSION_IP_KEY] = session_ip
        self.request.META["REMOTE_ADDR"] = valid
        self.assertTrue(self.middleware.process_request(self.request) is None)

        if invalid:
            self.request.session["canary"] = "canary"
            self.request.META["REMOTE_ADDR"] = invalid
            self.assertEqual(self.middleware.process_request(self.request).status_code, 400)
            self.assertFalse("canary" in self.request.session)

    def test_ip_was_known_now_absent(self):
        # Given: A session with a known ip
        session_ip = "127.0.0.1"
        self.add_session_to_request()
        self.request.session[middleware.SESSION_IP_KEY] = session_ip
        # When: Incoming request doesn't have Remote address
        self.request.META["REMOTE_ADDR"] = None
        response = self.middleware.process_request(self.request)
        # Then: there was an HttpResponse returned from middleware
        self.assertIsInstance(response, HttpResponse)
        # Then: the response defaults to 400 error
        self.assertEqual(response.status_code, 400)

    def test_validates_ua(self):
        self.add_session_to_request()

        self.request.META["HTTP_USER_AGENT"] = "test-ua1"
        self.request.session[middleware.SESSION_UA_KEY] = "test-ua1"
        self.assertTrue(self.middleware.process_request(self.request) is None)

        self.request.META["HTTP_USER_AGENT"] = "test-ua2"
        self.assertEqual(self.middleware.process_request(self.request).status_code, 400)

    @override_settings(RESTRICTEDSESSIONS_RESTRICT_IP=False)
    def test_disable_ip_validation(self):
        self._remote_addr_test(session_ip="127.0.0.1", valid="192.0.2.0")

    @override_settings(RESTRICTEDSESSIONS_RESTRICT_UA=False)
    def test_disable_ua_validation(self):
        self.add_session_to_request()
        self.request.session[middleware.SESSION_UA_KEY] = "test-ua1"
        self.request.META["HTTP_USER_AGENT"] = "test-ua2"
        self.assertTrue(self.middleware.process_request(self.request) is None)

    @override_settings(RESTRICTEDSESSIONS_AUTHED_ONLY=True)
    def test_only_authed_users_setting(self):
        self.add_session_to_request()
        self.request.user = AnonymousUser()
        self.assertIsNone(self.middleware.process_request(self.request))

        self.request.user = User(username="test")
        self.assertIsNone(self.middleware.process_request(self.request))
        self._remote_addr_test(session_ip="127.0.0.1", valid="127.0.0.1", invalid="127.0.0.2")
        self.assertIsInstance(self.request.user, AnonymousUser)

    def add_session_to_request(self):
        middleware = SessionMiddleware(self.get_response)
        middleware.process_request(self.request)
        self.request.session.save()
        # Trigger saving the session
        self.request.session["foo"] = "foo"


runtests.py
import sys

import django
from django.core.management import call_command

try:
    from django.conf import settings

    settings.configure(
        DEBUG=True,
        USE_TZ=True,
        SECRET_KEY='VERY_SECRET',
        DATABASES={
            "default": {
                "NAME": "./test.sqlite",
                "ENGINE": "django.db.backends.sqlite3",
            }
        },
        ROOT_URLCONF="test_urls",
        INSTALLED_APPS=[
            "django.contrib.auth",
            "django.contrib.contenttypes",
            "django.contrib.sites",
            "django.contrib.sessions",
            "restrictedsessions",
        ],
        MIDDLEWARE=[
            'django.contrib.sessions.middleware.SessionMiddleware',
            'restrictedsessions.middleware.RestrictedSessionsMiddleware',
            'django.middleware.common.CommonMiddleware',
            'django.middleware.csrf.CsrfViewMiddleware',
            'django.contrib.auth.middleware.AuthenticationMiddleware',
            'django.contrib.messages.middleware.MessageMiddleware',
            'django.middleware.clickjacking.XFrameOptionsMiddleware',
        ],
        SITE_ID=1,
    )
    django.setup()

except ImportError:
    raise ImportError("To fix this error, run: pip install -r requirements-test.txt")


def run_tests(*test_args):
    # Run tests
    call_command('test', test_args)


if __name__ == '__main__':
    run_tests(*sys.argv[1:])


restrictedsessions/middleware.py
# -*- coding: utf-8 -*-
from netaddr import IPNetwork, IPAddress, AddrConversionError, AddrFormatError
import logging

from django.conf import settings
from django.http import HttpResponse
from django.urls import reverse
from django.shortcuts import redirect

try:
    from django.utils.deprecation import MiddlewareMixin
except ImportError:
    MiddlewareMixin = object


SESSION_IP_KEY = "_restrictedsessions_ip"
SESSION_UA_KEY = "_restrictedsessions_ua"

logger = logging.getLogger("restrictedsessions")


class RestrictedSessionsMiddleware(MiddlewareMixin):
    def process_request(self, request):
        # Short circuit when request doesn't have session
        if not hasattr(request, "session"):
            return

        # Short circuit for option to require authed users
        if getattr(settings, "RESTRICTEDSESSIONS_AUTHED_ONLY", False):
            user = getattr(request, "user", None)
            # No logged in user -- ignore checks
            if not user or not hasattr(user, "is_authenticated") or not user.is_authenticated:
                return

        # Extract remote IP address for validation purposes
        remote_addr_key = getattr(settings, "RESTRICTEDSESSIONS_REMOTE_ADDR_KEY", "REMOTE_ADDR")
        remote_addr = request.META.get(remote_addr_key)
        if remote_addr:
            remote_addr = remote_addr.split(",")[0]

        # Clear the session and handle response when request doesn't validate
        if not self.validate_ip(request, remote_addr) or not self.validate_ua(request):
            if getattr(settings, "RESTRICTEDSESSIONS_AUTHED_ONLY", False):
                from django.contrib.auth import logout

                logout(request)
            else:  # logout(...) flushes the session so ensure it only happens once
                request.session.flush()
            logger.warning("Destroyed session due to invalid change of remote host or user agent")
            redirect_view = getattr(settings, "RESTRICTEDSESSIONS_REDIRECT_VIEW", None)
            if redirect_view:
                return redirect(reverse(redirect_view))
            else:
                status = getattr(settings, "RESTRICTEDSESSIONS_FAILURE_STATUS", 400)
                return HttpResponse(status=status)

        # Set the UA/IP Address on the session since they validated correctly
        request.session[SESSION_IP_KEY] = remote_addr
        if request.META.get("HTTP_USER_AGENT"):
            request.session[SESSION_UA_KEY] = request.META["HTTP_USER_AGENT"]

    def validate_ip(self, request, remote_ip):
        # When we aren't configured to restrict on IP address
        if not getattr(settings, "RESTRICTEDSESSIONS_RESTRICT_IP", True):
            return True
        # When the IP address key hasn't yet been set on the request session
        if SESSION_IP_KEY not in request.session:
            return True
        # When there is no remote IP, check if one has been set on the session
        session_ip = request.session[SESSION_IP_KEY]
        if not remote_ip:
            if session_ip:  # session has remote IP value so validate :-(
                return False
            else:  # Session doesn't have remote IP value so possibly :-)
                return True

        # Compute fuzzy IP compare based on settings on compare sensitivity
        session_network = IPNetwork(session_ip)
        remote_ip = IPAddress(remote_ip)
        try:
            session_network = session_network.ipv4()
            remote_ip = remote_ip.ipv4()
            session_network.prefixlen = getattr(settings, "RESTRICTEDSESSIONS_IPV4_LENGTH", 32)
        except AddrConversionError:
            try:
                session_network.prefixlen = getattr(settings, "RESTRICTEDSESSIONS_IPV6_LENGTH", 64)
            except AddrFormatError:
                # session_network must be IPv4, but remote_ip is IPv6
                return False
        return remote_ip in session_network

    def validate_ua(self, request):
        # When we aren't configured to restrict on user agent
        if not getattr(settings, "RESTRICTEDSESSIONS_RESTRICT_UA", True):
            return True
        # When the user agent key hasn't been set yet in the request session
        if SESSION_UA_KEY not in request.session:
            return True
        # Compare the new user agent value with what is known about the session
        return request.session[SESSION_UA_KEY] == request.META["HTTP_USER_AGENT"]
